Nostr Double Ratchet

mmalmi opened the pull request:
Implementations:
* [nostr-double-ratchet](https://github.com/mmalmi/nostr-double-ratchet) typescript library, deployed at [iris.to](https://iris.to)

mmalmi committed:
double ratchet

alltheseas commented:
@mmalmi does this leave the door open for interoperability with signal messenger?

mmalmi commented:
> @mmalmi does this leave the door open for interoperability with signal messenger?

Unfortunately not, unless we get them on Nostr 😄 This uses rotated Nostr public keys for communication and NIP-44 encryption.

fiatjaf commented:
> @mmalmi does this leave the door open for interoperability with signal messenger?

If Signal wanted interoperability they wouldn't ban alternative clients to their API or force people to use their central server.

1l0 commented on 117.md:
````diff
@@ -0,0 +1,366 @@
+```typescript
+sendEvent(event: Partial<UnsignedEvent>) {
+  const innerEvent = nip59.createRumor(event)
+  const [header, encryptedData] = this.ratchetEncrypt(JSON.stringify(rumor));
````
```suggestion
  const [header, encryptedData] = this.ratchetEncrypt(JSON.stringify(innerEvent));
```

mmalmi committed:
Update 117.md

Co-authored-by: reis <1l0@users.noreply.github.com>

1l0 commented:
You're listing "group channel" in your repo—does that mean you plan to implement it?

mmalmi committed:
better initialization code

mmalmi committed:
add recommendation to use recipient inbox relays

TheAwiteb commented on 117.md:
```diff
@@ -0,0 +1,361 @@
+
+#### Inner event
+
+Inner events must be [NIP-59](https://github.com/nostr-protocol/nips/blob/master/59.md) Rumors (unsigned Nostr events) allowing plausible deniability.
```
```suggestion
Inner events must be [NIP-59](59.md) Rumors (unsigned Nostr events) allowing plausible deniability.
```

TheAwiteb commented on 117.md:
```diff
@@ -0,0 +1,361 @@
+
+### External functions
+
+We use the following Nostr functions ([NIP-01](https://github.com/nostr-protocol/nips/blob/master/01.md)):
```
```suggestion
We use the following Nostr functions ([NIP-01](01.md)):
```

TheAwiteb commented on 117.md:
```diff
@@ -0,0 +1,361 @@
+- `generateSecretKey()` for creating Nostr private keys
+- `finalizeEvent(partialEvent, secretKey)` for creating valid Nostr events with pubkey, id and signature
+
+We use [NIP-44](https://github.com/nostr-protocol/nips/blob/master/44.md) functions for encryption:
```
```suggestion
We use [NIP-44](44.md) functions for encryption:
```

TheAwiteb commented on 117.md:
```diff
@@ -0,0 +1,361 @@
+- `nip44.decrypt`
+- `nip44.getConversationKey`
+
+[NIP-59](https://github.com/nostr-protocol/nips/blob/master/59.md):
```
```suggestion
[NIP-59](59.md):
```

TheAwiteb commented on 118.md:
````diff
@@ -0,0 +1,86 @@
+}
+```
+
+`l` is a [NIP-32](https://github.com/nostr-protocol/nips/blob/master/32.md) label tag that can be used to list your invites. We use this approach with [NIP-78](./78.md) application specific data kind `30078` in order to not bloat the event kind space and instead have a human-readable label.
````
```suggestion
`l` is a [NIP-32](32.md) label tag that can be used to list your invites. We use this approach with [NIP-78](./78.md) application specific data kind `30078` in order to not bloat the event kind space and instead have a human-readable label.
```

TheAwiteb commented on 118.md:
```diff
@@ -0,0 +1,86 @@
+
+#### Outer event
+
+Invite response outer event is a [NIP-59](https://github.com/nostr-protocol/nips/blob/master/59.md) gift wrap event of kind `1059`, sent from a random, one-time-use pubkey, hiding the responder's identity from the public.
```
```suggestion
Invite response outer event is a [NIP-59](59.md) gift wrap event of kind `1059`, sent from a random, one-time-use pubkey, hiding the responder's identity from the public.
```

TheAwiteb commented on 118.md:
```diff
@@ -0,0 +1,86 @@
+
+#### Inner event
+
+[NIP-59](https://github.com/nostr-protocol/nips/blob/master/59.md) Rumor.
```
```suggestion
[NIP-59](59.md) Rumor.
```

mmalmi committed:
Update 117.md

Co-authored-by: Awiteb <a@4rs.nl>

mmalmi committed:
Update 117.md

Co-authored-by: Awiteb <a@4rs.nl>

mmalmi committed:
Update 117.md

Co-authored-by: Awiteb <a@4rs.nl>

mmalmi committed:
Update 117.md

Co-authored-by: Awiteb <a@4rs.nl>

mmalmi committed:
Update 117.md

Co-authored-by: Awiteb <a@4rs.nl>

mmalmi committed:
Update 118.md

Co-authored-by: Awiteb <a@4rs.nl>

mmalmi committed:
Update 118.md

Co-authored-by: Awiteb <a@4rs.nl>

mmalmi committed:
Update 118.md

Co-authored-by: Awiteb <a@4rs.nl>

vitorpamplona commented on 117.md:
```diff
@@ -0,0 +1,361 @@
+
+The encrypted header contains our next nostr public key, our previous sending chain length and the current message number.
+
+The events should be sent to the recipient's inbox relays.
```
Shouldn't this use the DM relays? Otherwise, if I am the only user of my public inbox, it's possible to identify DMs sent to me there.

vitorpamplona commented on 117.md:
```diff
@@ -0,0 +1,361 @@
+
+The Double Ratchet is a key rotation algorithm for secure private messaging.
+
+It allows us to 1) communicate on Nostr without revealing metadata (who you are communicating with and when), and 2) keep your message history and future messages safe even if your main Nostr key is compromised.
```
```suggestion
It allows us to 1) communicate on Nostr without revealing the receiver of each message, and 2) keep your message history and future messages safe even if your main Nostr key is compromised.
```
NIP-17 already hides the sender and the date/time

vitorpamplona commented on 117.md:
```diff
@@ -0,0 +1,361 @@
+  kind: 1060,
+  content: encryptedInnerEvent,
+  tags: [["header", encryptedHeader]],
+  pubkey: ratchetPublicKey,
```
Which ratchet is this key from, 1 or 2?

vitorpamplona commented on 117.md:
```diff
@@ -0,0 +1,361 @@
+
+We subscribe to Double Ratchet events based on author public keys which are ephemeral — not used for other purposes than the Double Ratchet session. We use the regular event kind `1060` to differentiate it from other DM kinds, retrieval of which may be restricted by relays.
+
+The encrypted header contains our next nostr public key, our previous sending chain length and the current message number.
```
What happens if users miss a message (like the relay deleted it or the sending process failed) and you don't know what is the next pubkey?

vitorpamplona commented on 117.md:
```diff
@@ -0,0 +1,361 @@
+
+With established Nostr event kinds, clients can implement all kinds of features, such as replies, reactions, and encrypted file sharing in private messages.
+
+Direct message and encrypted file messages are defined in [NIP-17](17.md).
```
Does it allow groups with the multiple p-tags from NIP-17?

vitorpamplona commented on 117.md:
```diff
@@ -0,0 +1,361 @@
+  content: encryptedInnerEvent,
+  tags: [["header", encryptedHeader]],
+  pubkey: ratchetPublicKey,
+  created_at,
```
Does this event also randomizes time to avoid time-collision attacks, like on NIP-17?

Otherwise, couldn't an attacker figure out the two keys talking to one another by looking at the create_at sets?

vitorpamplona commented on 118.md:
```diff
@@ -0,0 +1,86 @@
+    ['d', `double-ratchet/invites/${uid}`], // uid: unique name for the invitation, e.g. "public"
+    ['l', 'double-ratchet/invites'], // label, for listing all your invites
+    ['ephemeralKey', inviterEphemeralKey],
+    ['sharedSecret', sharedSecret]
```
Is the shared secret public?

vitorpamplona commented on 118.md:
````diff
@@ -0,0 +1,86 @@
+}
+```
+
+`l` is a [NIP-32](32.md) label tag that can be used to list your invites. We use this approach with [NIP-78](./78.md) application specific data kind `30078` in order to not bloat the event kind space and instead have a human-readable label.
````
If this is for other folks, it should probably not use 30078 and create it's own kind.

vitorpamplona commented on 118.md:
```diff
@@ -0,0 +1,86 @@
+  tags: [
+    ['d', `double-ratchet/invites/${uid}`], // uid: unique name for the invitation, e.g. "public"
+    ['l', 'double-ratchet/invites'], // label, for listing all your invites
+    ['ephemeralKey', inviterEphemeralKey],
```
Is this a public key or private key?

vitorpamplona commented on 117.md:
```diff
@@ -0,0 +1,361 @@
+
+We implement the Double Ratchet Algorithm on Nostr similarly to Signal's [Double Ratchet with header encryption](https://signal.org/docs/specifications/doubleratchet/#double-ratchet-with-header-encryption), but encrypting the message headers with [NIP-44](./44.md) conversation keys instead of symmetric header keys.
+
+Ratchet 1 keys are standard Nostr keys. In addition to encryption, they are also used for publishing and subscribing to messages on Nostr. As they are rotated and not linked to public Nostr identities, metadata privacy is preserved.
```
What do you mean by `subscripting to`? Which keys should be used to download kind 1060's from the relay? Should clients just list all rotated keys from the ratchet to download each message? Doesn't this let the relay know all the keys for a given section? Also, should we AUTH into relays? Because that would connect the ratchet keys with the nostr pubkey of the user and from there all the metadata can be found. The relay just need to publish keys they have seen from time to time.

vitorpamplona reviewed (commented):
Lots of questions :)

Commit:
NIP-117/118: align drafts with interoperable protocol

Correct the initial DH key, multi-ratchet skipped-key routing, validation and atomicity requirements, and the deployed direct kind-1059 invite response construction.

Commit:
Clarify Double Ratchet security requirements

Require exact unsigned rumors, authenticated and rollback-safe session persistence, and a session-key proof of possession in invite responses.\n\nThe proof binds both identities, the inviter ephemeral key, the session key, and the shared secret to prevent unknown-key-share identity misbinding.
